pkg/controller/cloudprivateipconfig/cloudprivateipconfig_controller.go

```diff
@@ -0,0 +1,506 @@
+				controllerutil.ContainsFinalizer(newCloudPrivateIPConfig, cloudPrivateIPConfigFinalizer)) ||
+				!reflect.DeepEqual(oldCloudPrivateIPConfig.Spec, newCloudPrivateIPConfig.Spec) {
+				controller.Enqueue(new)
+			}
+		},
+		DeleteFunc: controller.Enqueue,
+	})
+	return controller
+}
+
+// syncHandler compares the actual state with the desired, and attempts to
+// converge the two. It then updates the Status block of the CloudPrivateIPConfig
+// resource with the current status of the resource.
+// On update: we should only process the add once we've received the cloud's answer
+// for the delete. We risk having the IP address being assigned to two nodes at
+// the same time otherwise.
+
+// We have two "data stores": kube API server and the cloud API. Thus, there are
+// two main error conditions:
+
+// - We couldn't update the object in the kube API, but did update the object in the cloud.
+// - We couldn't update the object in the cloud, but did update the object in the kube API.
+
+// - If we couldn't update either, we just resync the original object
+// - If we could update both, we don't resync the object
+
+//  Here's a schema of CloudPrivateIPConfig's reconciliation loop based on the consumer input:
+
+// - ADD:
+// 1. 	Send cloud API ADD request
+// 2. 	Set status.node = spec.node && status.conditions[0].Status = Pending
+// ...some time later
+// 3.	Get cloud API ADD result
+// * 	If OK: set status.conditions[0].Status = Success
+// *	If !OK: set status.node == "" && set status.conditions[0].Status = Error && goto 1. by resync
+
+// - DELETE:
+// 1.	Send cloud API DELETE request
+// 2. 	Set status.conditions[0].Status = Pending (keep old status.node)
+// ...some time later
+// 3. 	Get cloud API DELETE result
+// *	If OK: -
+// * 	If !OK: set status.conditions[0].Status = Error && goto 1. by resync
+
+// - UPDATE:
+// 1.	goto DELETE
+// *	If OK: status.node = "" && goto ADD
+
+// Consumer should only consider ADD / UPDATE successful when:
+// - 	spec.node == status.node && status.conditions[0].Status == Success
+func (c *CloudPrivateIPConfigController) SyncHandler(key string) error {
+	var status *cloudnetworkv1.CloudPrivateIPConfigStatus
+	var cloudRequestObj interface{}
+	// Convert the key into a distinct name (since CloudPrivateIPConfig is
+	// cluster-scoped)
+	_, name, err := cache.SplitMetaNamespaceKey(key)
+	if err != nil {
+		utilruntime.HandleError(fmt.Errorf("invalid resource key: %s", key))
+		return nil
+	}
+
+	// This object will recursively be updated during this sync
+	cloudPrivateIPConfig, err := c.cloudPrivateIPConfigLister.Get(name)
+	if err != nil {
+		if apierrors.IsNotFound(err) {
+			// The object was deleted while we were processing the request,
+			// nothing more to do, the finalizer portion of this sync should
+			// have handled the last cleanup
+			utilruntime.HandleError(fmt.Errorf("CloudPrivateIPConfig: '%s' in work queue no longer exists", name))
```

Comment:
I mean, if it was going to be a best practice then we should add a version where you can just pass the message directly rather than having to create an error just so you can pass it to the function.

But `HandleError` is only interesting if you also use the `utilruntime` functions for overriding its behavior anyway. If you never call those, then it's just a long-winded way of calling `klog.Errorf`.